<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: World Development Statistics

--- 
# Part 1

Part 1 requires knowledge of basic Python.

---

## Problem Statement

Decide on your problem statement that will guide your analysis for this project. For guidelines, sample prompts, or inspiration, check out the README.

**To-Do:** *Replace this cell with your problem statement.*

Doctors-Go-Anywhere have reached out after their recent funding, as they are looking to expand their operations. They are wanting to indentify the Top 5 regions globally that have the least Healthcare access and would benefit from their services.

### Contents:
- [Background](#Background)
- [Data Import & Cleaning](#Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-Data)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

## Background

For our first project, we're going to take a look at World Development Statistics from [Gapminder](https://www.gapminder.org/about/), an independent Swedish foundation that aims to make data about the world more accessible and reliable. A good introduction on Gapminder is this [Ted Talk](https://www.ted.com/talks/hans_rosling_the_best_stats_you_ve_ever_seen) from Hans Rosling, which also shows how effective data visualization can be for your audience.

**To-Do:** *Fill out this cell (or edit the above cell) with any other background or information that is necessary for your problem statement.*

Additional will also be pulled from world health organization. We will also be looking into see the number of existing doctors that there are in regions and see how much governments are investing into their healthcare to see how that collerates into in their position on the rankings.


### Choose your Data

There are 3 datasets included in the [`data`](./data/) folder for this project. You are required to pick **at least two** of these to complete your analysis. Feel free to use more than two if you would like, or add other relevant datasets you find online. 

**HINT: There are more datasets available on the Gapminder website**

* [`population.csv`](./data/population.csv): Population by Country
* [`life_expectancy.csv`](./data/life_expectancy.csv): Life Expectancy by Country
* [`gni_per_cap_atlas_method_con2021.csv`](./data/gni_per_cap_atlas_method_con2021.csv): Gross National Income (GNI) per capita in current US dollars

**To-Do:** *Fill out this cell with the datasets you will use for your analysis. Write a brief description of the contents for each dataset that you choose.*

Alongside the population & life expectancy data files, I will be pulling data on the number of doctors, health aid given and total health spending.


### Outside Research

Based on your problem statement and your chosen datasets, spend some time doing outside research on state policies or additional information that might be relevant. Summarize your findings below. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. **Make sure that you cite your sources.**

**To-Do:** *Fill out this cell with outside research or any additional background information that will support your analysis.*

### Coding Challenges

1. Manually calculate mean:

    Write a function that takes in values and returns the mean of the values. Create a list of numbers that you test on your function to check to make sure your function works!
    
    *Note*: Do not use any mean methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [6]:
# Code:
list_numbers = [56, 577, 343, 332, 566, 21, 96]
mean = sum(list_numbers)/len(list_numbers); 

2. Manually calculate standard deviation:

    The formula for standard deviation is below:

    $$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

    Where $x_i$ represents each value in the dataset, $\mu$ represents the mean of all values in the dataset and $n$ represents the number of values in the dataset.

    Write a function that takes in values and returns the standard deviation of the values using the formula above. Hint: use the function you wrote above to calculate the mean! Use the list of numbers you created above to test on your function.
    
    *Note*: Do not use any standard deviation methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [7]:
# Code:
((sum((num - mean)**2 for num in list_numbers))/len(list_numbers)) ** 0.5


216.35081111331144

3. Data cleaning function:
    
    Write a function that takes in a string consisting a number and an abbreviated place value (ex. '5M', '45K'), and converts that string to an integer. For example, inputting '5M' in your function should return 5000000, '45K' should return 45000, etc. Make sure to test your function to make sure it works!

You will use these functions later on in the project!

In [8]:
# Code:
def string_tonum(org_fig):
    keys = {'K': 1000, 'M': 1000000, 'B': 1000000000, '0': 1}

    num = float(org_fig[:-1])
    letter = org_fig[-1].upper()
    
    if letter in keys:
        return num * keys[letter]
    
print(string_tonum('5M'))

5000000.0


In [9]:
popul23[popul23['2005'].astype(str).apply(lambda s: s[-1]).isin(['n', '0'])]

NameError: name 'popul23' is not defined

--- 
# Part 2

Part 2 requires knowledge of Pandas, EDA, data cleaning, and data visualization.

---

*All libraries used should be added here*

In [12]:
# Imports:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns



In [32]:
#Data files being used

population = pd.read_csv('../data/population.csv')
life_expect = pd.read_csv('../data/life_expectancy.csv')
health_aid = pd.read_csv('../data/API_SH.XPD.GHED.PP.CD_DS2_en_csv_v2_5872088.csv')
medical_doc = pd.read_csv('../data/medical_doctors_per_1000_people.csv')

## Data Import and Cleaning

### Data Import & Cleaning

Import the datasets that you selected for this project and go through the following steps at a minimum. You are welcome to do further cleaning as you feel necessary:
1. Display the data: print the first 5 rows of each dataframe to your Jupyter notebook.
2. Check for missing values.
3. Check for any obvious issues with the observations (keep in mind reasonable values for each column).
4. Fix any errors you identified in steps 2-3.
5. Display the data types of each feature.
6. Fix any incorrect data types found in step 5.
    - Fix any individual values preventing other columns from being the appropriate type.
    - If your dataset has a column of abbreviated numbers (ex. '3M', '45K', etc.), use the function you wrote in Part 1 (coding challenges, number 3) to convert this to integers! *Hint*: use `.map()` or `.apply()`.
7. Rename Columns.
    - Column names should be all lowercase.
    - Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`).
    - Column names should be unique and informative.
8. Drop unnecessary rows (if needed).
9. Merge dataframes that can be merged.
10. Perform any additional cleaning that you feel is necessary.
11. Save your cleaned and merged dataframes as csv files.

In [19]:
# Code population cleaning
population.head()
popul23 = population[['country'] + [str(year) for year in range(2000, 2024)]]
missing_pop = popul23[popul23.isnull().any(axis=1)]
print(missing_pop)



Empty DataFrame
Columns: [country, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023]
Index: []

[0 rows x 25 columns]


In [15]:
#removing Holy See from dataset as too many dates points missing
population = population.dropna() #dropping Holy See for a lot of missing data
popul23.head()

country   2000   2001   2002   2003   2004   2005   2006  \
0           Afghanistan  19.5M  19.7M    21M  22.6M  23.6M  24.4M  25.4M   
1                Angola  16.4M  16.9M  17.5M  18.1M  18.8M  19.5M  20.2M   
2               Albania  3.18M  3.15M  3.12M  3.09M  3.06M  3.03M     3M   
3               Andorra  66.1k  67.8k  70.8k  73.9k  76.9k  79.8k  80.2k   
4  United Arab Emirates  3.28M  3.45M  3.63M  3.81M  3.99M  4.28M   4.9M   

    2007   2008  ...   2014   2015   2016   2017   2018   2019   2020   2021  \
0  25.9M  26.4M  ...  32.7M  33.8M  34.6M  35.6M  36.7M  37.8M    39M  40.1M   
1  20.9M  21.7M  ...  27.1M  28.1M  29.2M  30.2M  31.3M  32.4M  33.4M  34.5M   
2  2.98M  2.95M  ...  2.88M  2.88M  2.88M  2.88M  2.88M  2.87M  2.87M  2.85M   
3  78.2k  76.1k  ...  71.6k  71.7k  72.5k  73.8k    75k  76.3k  77.7k    79k   
4  5.87M  6.99M  ...  8.84M  8.92M  8.99M  9.07M  9.14M  9.21M  9.29M  9.37M   

    2022   2023  
0  41.1M  42.2M  
1  35.6M  36.7M  
2  2.84M  2.83M  
3  79.8k  80.1k  
4  9.44M  9.52M  

[5 rows x 25 columns]

In [16]:
popul23.dtypes

country    object
2000       object
2001       object
2002       object
2003       object
2004       object
2005       object
2006       object
2007       object
2008       object
2009       object
2010       object
2011       object
2012       object
2013       object
2014       object
2015       object
2016       object
2017       object
2018       object
2019       object
2020       object
2021       object
2022       object
2023       object
dtype: object

In [20]:
#population = population.drop(74) #74 for the row that Holy see was on
#running missing statement again to ensure it has been removed
#popul23.isnull().sum()

#Converting the numbers from abbreviations
for yrs in popul23.columns[1:]:
    popul23[yrs] = popul23[yrs].map(string_tonum)

#popul23.iloc[:, 1:] = popul23.iloc[:, 1:].apply(string_tonum)   

/var/folders/pj/jq7g3ftn5557342n7bvxr3q80000gn/T/ipykernel_37187/329833947.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  popul23[yrs] = popul23[yrs].map(string_tonum)


In [21]:
popul23

country        2000        2001        2002        2003  \
0             Afghanistan  19500000.0  19700000.0  21000000.0  22600000.0   
1                  Angola  16400000.0  16900000.0  17500000.0  18100000.0   
2                 Albania   3180000.0   3150000.0   3120000.0   3090000.0   
3                 Andorra     66100.0     67800.0     70800.0     73900.0   
4    United Arab Emirates   3280000.0   3450000.0   3630000.0   3810000.0   
..                    ...         ...         ...         ...         ...   
192                 Samoa    184000.0    186000.0    187000.0    187000.0   
193                 Yemen  18600000.0  19100000.0  19700000.0  20200000.0   
194          South Africa  46800000.0  47200000.0  47700000.0  48100000.0   
195                Zambia   9890000.0  10200000.0  10500000.0  10800000.0   
196              Zimbabwe  11800000.0  11900000.0  12000000.0  12100000.0   

           2004        2005        2006        2007        2008  ...  \
0    23600000.0  24400000.0  25400000.0  25900000.0  26400000.0  ...   
1    18800000.0  19500000.0  20200000.0  20900000.0  21700000.0  ...   
2     3060000.0   3030000.0   3000000.0   2980000.0   2950000.0  ...   
3       76900.0     79800.0     80200.0     78200.0     76100.0  ...   
4     3990000.0   4280000.0   4900000.0   5870000.0   6990000.0  ...   
..          ...         ...         ...         ...         ...  ...   
192    188000.0    189000.0    189000.0    190000.0    192000.0  ...   
193  20700000.0  21300000.0  22000000.0  22600000.0  23300000.0  ...   
194  48600000.0  49000000.0  49500000.0  50000000.0  50600000.0  ...   
195  11200000.0  11600000.0  12000000.0  12400000.0  12900000.0  ...   
196  12200000.0  12200000.0  12300000.0  12500000.0  12600000.0  ...   

           2014        2015        2016        2017        2018        2019  \
0    32700000.0  33800000.0  34600000.0  35600000.0  36700000.0  37800000.0   
1    27100000.0  28100000.0  29200000.0  30200000.0  31300000.0  32400000.0   
2     2880000.0   2880000.0   2880000.0   2880000.0   2880000.0   2870000.0   
3       71600.0     71700.0     72500.0     73800.0     75000.0     76300.0   
4     8840000.0   8920000.0   8990000.0   9070000.0   9140000.0   9210000.0   
..          ...         ...         ...         ...         ...         ...   
192    202000.0    204000.0    206000.0    208000.0    210000.0    212000.0   
193  27800000.0  28500000.0  29300000.0  30000000.0  30800000.0  31500000.0   
194  54700000.0  55900000.0  56400000.0  56600000.0  57300000.0  58100000.0   
195  15700000.0  16200000.0  16800000.0  17300000.0  17800000.0  18400000.0   
196  13900000.0  14200000.0  14500000.0  14800000.0  15100000.0  15400000.0   

           2020        2021        2022        2023  
0    39000000.0  40100000.0  41100000.0  42200000.0  
1    33400000.0  34500000.0  35600000.0  36700000.0  
2     2870000.0   2850000.0   2840000.0   2830000.0  
3       77700.0     79000.0     79800.0     80100.0  
4     9290000.0   9370000.0   9440000.0   9520000.0  
..          ...         ...         ...         ...  
192    215000.0    219000.0    222000.0    226000.0  
193  32300000.0  33000000.0  33700000.0  34400000.0  
194  58800000.0  59400000.0  59900000.0  60400000.0  
195  18900000.0  19500000.0  20000000.0  20600000.0  
196  15700000.0  16000000.0  16300000.0  16700000.0  

[196 rows x 25 columns]

In [ ]:
#Saving the cleaned population file
popul23.to_csv('Population23.csv', index = False)

In [22]:
#Life Expectancy cleaning
life_expect.head()


country  1800  1801  1802  1803  1804  1805  1806  1807  1808  \
0           Afghanistan  28.2  28.2  28.2  28.2  28.2  28.2  28.1  28.1  28.1   
1                Angola  27.0  27.0  27.0  27.0  27.0  27.0  27.0  27.0  27.0   
2               Albania  35.4  35.4  35.4  35.4  35.4  35.4  35.4  35.4  35.4   
3               Andorra   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  United Arab Emirates  30.7  30.7  30.7  30.7  30.7  30.7  30.7  30.7  30.7   

   ...  2091  2092  2093  2094  2095  2096  2097  2098  2099  2100  
0  ...  75.5  75.7  75.8  76.0  76.1  76.2  76.4  76.5  76.6  76.8  
1  ...  78.8  79.0  79.1  79.2  79.3  79.5  79.6  79.7  79.9  80.0  
2  ...  87.4  87.5  87.6  87.7  87.8  87.9  88.0  88.2  88.3  88.4  
3  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4  ...  82.4  82.5  82.6  82.7  82.8  82.9  83.0  83.1  83.2  83.3  

[5 rows x 302 columns]

In [23]:
#creating new variable with only 2000 to 2023
life_expect23 = life_expect[['country'] + [str(year) for year in range(2000, 2024)]]

In [24]:
#locating missing data
life_expect23[life_expect23.isnull().any(axis=1)]

country  2000  2001  2002  2003  2004  2005  2006  2007  \
3                Andorra  80.8  80.9  81.1  81.2  81.3  81.4  81.5  81.7   
46              Dominica  72.1  72.2  72.3  72.2  72.3  72.6  72.6  72.5   
93   St. Kitts and Nevis  69.9  70.3  70.7  71.1  71.2  71.4  71.5  71.7   
107               Monaco  78.8  78.9  78.9  79.0  79.0  79.0  79.0  79.1   
112     Marshall Islands  62.7  63.0  63.0  63.0  63.1  63.1  63.2  63.4   
131                Nauru  59.1  58.9  58.8  58.7  58.7  58.8  58.9  59.0   
138                Palau  66.4  66.5  66.7  66.8  66.8  66.8  66.9  67.0   
156           San Marino  81.5  81.6  81.7  81.7  81.6  81.6  81.6  81.7   
178               Tuvalu  61.5  64.7  65.1  65.3  65.5  65.7  66.0  66.1   

     2008  ...  2014  2015  2016  2017  2018  2019  2020  2021  2022  2023  
3    81.8  ...  82.0  82.0  82.1  82.1  82.1  82.2   NaN   NaN   NaN   NaN  
46   72.7  ...  72.2  71.5  72.3  71.2  72.3  72.3   NaN   NaN   NaN   NaN  
93   72.0  ...  72.1  72.0  72.1  72.1  72.1  72.1   NaN   NaN   NaN   NaN  
107  79.1  ...  79.4  79.5  79.7  79.8  79.9  80.1   NaN   NaN   NaN   NaN  
112  63.6  ...  64.8  64.9  65.1  65.2  65.3  65.5   NaN   NaN   NaN   NaN  
131  59.3  ...  61.9  62.3  62.6  63.0  63.2  63.4   NaN   NaN   NaN   NaN  
138  67.0  ...  67.1  67.2  67.3  67.4  67.4  67.5   NaN   NaN   NaN   NaN  
156  81.8  ...  82.0  82.0  82.0  82.0  82.1  82.2   NaN   NaN   NaN   NaN  
178  66.3  ...  67.2  67.4  67.5  67.7  67.8  67.9   NaN   NaN   NaN   NaN  

[9 rows x 25 columns]

In [25]:
#Due to the missing data above dropping years 2020-2023
life_expect19 = life_expect23[['country'] + [str(year) for year in range(2000, 2020)]]
life_expect19.head()

country  2000  2001  2002  2003  2004  2005  2006  2007  2008  \
0           Afghanistan  54.7  54.8  55.5  56.5  57.1  57.6  58.0  58.5  59.2   
1                Angola  52.8  53.4  54.5  55.1  55.5  56.4  57.0  58.0  58.8   
2               Albania  75.4  76.0  75.9  75.6  75.8  76.2  76.9  77.5  77.6   
3               Andorra  80.8  80.9  81.1  81.2  81.3  81.4  81.5  81.7  81.8   
4  United Arab Emirates  69.1  69.2  69.4  69.3  69.1  69.2  69.5  70.0  70.4   

   ...  2010  2011  2012  2013  2014  2015  2016  2017  2018  2019  
0  ...  60.5  61.0  61.4  61.9  61.9  61.9  62.0  62.9  62.7  63.3  
1  ...  60.2  60.8  61.4  62.1  63.0  63.5  63.9  64.2  64.6  65.1  
2  ...  78.1  78.1  78.2  78.3  78.2  78.1  78.2  78.3  78.4  78.5  
3  ...  81.8  81.9  81.9  82.0  82.0  82.0  82.1  82.1  82.1  82.2  
4  ...  70.8  71.0  71.2  71.6  73.0  73.2  73.4  73.5  73.7  73.9  

[5 rows x 21 columns]

In [35]:
#Health aid data cleaning
health_aid.head()

Country Name  1960  1961  1962  1963  1964  1965  1966  \
0                        Aruba   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1  Africa Eastern and Southern   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2                  Afghanistan   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
3   Africa Western and Central   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4                       Angola   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   1967  1968  ...        2013        2014       2015       2016       2017  \
0   NaN   NaN  ...         NaN         NaN        NaN        NaN        NaN   
1   NaN   NaN  ...   87.532252   92.625452  98.537006  97.532557  94.368510   
2   NaN   NaN  ...    8.965855   10.032695  11.157826  12.145635  13.480682   
3   NaN   NaN  ...   29.627769   28.521052  31.641008  29.177849  30.135567   
4   NaN   NaN  ...  129.577242  107.074785  89.779783  84.152683  93.222057   

        2018       2019        2020  2021  2022  
0        NaN        NaN         NaN   NaN   NaN  
1  95.377785  98.107913  102.888434   NaN   NaN  
2  11.656513  10.775585   24.629420   NaN   NaN  
3  30.812392  32.635270   40.102642   NaN   NaN  
4  77.214999  70.747300   77.449551   NaN   NaN  

[5 rows x 64 columns]

In [36]:
health_aid20 = health_aid[['Country Name'] + [str(year) for year in range(2000,2020)]]
health_aid20

Country Name        2000        2001        2002  \
0                          Aruba         NaN         NaN         NaN   
1    Africa Eastern and Southern   43.326868   43.616394   43.292173   
2                    Afghanistan         NaN         NaN    0.779684   
3     Africa Western and Central   15.069751   19.874533   16.521509   
4                         Angola   36.498135   69.275662   49.542771   
..                           ...         ...         ...         ...   
261                       Kosovo         NaN         NaN         NaN   
262                  Yemen, Rep.   65.549442   70.729858   68.381964   
263                 South Africa  219.173157  204.060650  204.394199   
264                       Zambia   55.111193   56.189848   55.107062   
265                     Zimbabwe         NaN         NaN         NaN   

           2003        2004        2005        2006        2007        2008  \
0           NaN         NaN         NaN         NaN         NaN         NaN   
1     51.024088   53.532174   56.310836   66.328330   72.791742   75.928888   
2      6.277876    5.199076    5.629746    5.538590    6.339159    8.420790   
3     23.263107   29.129152   30.981240   28.343067   27.206899   26.781339   
4     56.252356   71.218851   61.807417   77.386219  104.169230  141.319476   
..          ...         ...         ...         ...         ...         ...   
261         NaN         NaN         NaN         NaN         NaN         NaN   
262   72.663043   61.366065   55.073121   55.844830   55.315939   58.119578   
263  257.033126  267.605490  289.189068  366.687199  401.232261  428.057922   
264   48.018202   50.593274   41.036901   41.521431   25.281655   15.888849   
265         NaN         NaN         NaN         NaN         NaN         NaN   

     ...        2010        2011        2012        2013        2014  \
0    ...         NaN         NaN         NaN         NaN         NaN   
1    ...   82.528924   85.130918   84.968293   87.532252   92.625452   
2    ...    8.143219    8.260166    6.746567    8.965855   10.032695   
3    ...   25.676549   28.060090   29.601457   29.627769   28.521052   
4    ...  110.251719  114.787540  115.519792  129.577242  107.074785   
..   ...         ...         ...         ...         ...         ...   
261  ...         NaN         NaN         NaN         NaN         NaN   
262  ...   45.269081   36.814275   40.444415   33.673440   26.624408   
263  ...  510.862585  547.265420  543.960807  562.069326  592.881947   
264  ...   26.385108   39.335791   43.578984   45.929328   65.591017   
265  ...   49.852982   44.102007   46.952384   50.896451   69.977044   

           2015        2016        2017        2018        2019  
0           NaN         NaN         NaN         NaN         NaN  
1     98.537006   97.532557   94.368510   95.377785   98.107913  
2     11.157826   12.145635   13.480682   11.656513   10.775585  
3     31.641008   29.177849   30.135567   30.812392   32.635270  
4     89.779783   84.152683   93.222057   77.214999   70.747300  
..          ...         ...         ...         ...         ...  
261         NaN         NaN         NaN         NaN         NaN  
262   10.415752         NaN         NaN         NaN         NaN  
263  624.261610  626.321901  636.913732  660.930514  693.963122  
264   70.085924   56.959218   60.753665   69.705245   75.097565  
265   40.569432   43.111729   56.813880   61.475349   18.839995  

[266 rows x 21 columns]

In [37]:
# Identifying all missing data

health_aid20[health_aid20.isnull().any(axis=1)]

Country Name         2000         2001         2002  \
0                        Aruba          NaN          NaN          NaN   
2                  Afghanistan          NaN          NaN     0.779684   
11              American Samoa          NaN          NaN          NaN   
27                     Bermuda          NaN          NaN          NaN   
38             Channel Islands          NaN          NaN          NaN   
51                     Curacao          NaN          NaN          NaN   
52              Cayman Islands          NaN          NaN          NaN   
78               Faroe Islands          NaN          NaN          NaN   
84                   Gibraltar          NaN          NaN          NaN   
91                   Greenland          NaN          NaN          NaN   
93                        Guam          NaN          NaN          NaN   
96        Hong Kong SAR, China          NaN          NaN          NaN   
108                Isle of Man          NaN          NaN          NaN   
110             Not classified          NaN          NaN          NaN   
113                       Iraq          NaN          NaN          NaN   
132                      Libya   312.915891   405.994888   591.069076   
137              Liechtenstein          NaN          NaN          NaN   
146           Macao SAR, China          NaN          NaN          NaN   
147   St. Martin (French part)          NaN          NaN          NaN   
162                 Montenegro          NaN          NaN          NaN   
164   Northern Mariana Islands          NaN          NaN          NaN   
172              New Caledonia          NaN          NaN          NaN   
192                Puerto Rico          NaN          NaN          NaN   
193  Korea, Dem. People's Rep.          NaN          NaN          NaN   
196         West Bank and Gaza          NaN          NaN          NaN   
199           French Polynesia          NaN          NaN          NaN   
205               Saudi Arabia  1160.633409  1219.260363  1115.614260   
213                    Somalia          NaN          NaN          NaN   
216                South Sudan          NaN          NaN          NaN   
225  Sint Maarten (Dutch part)          NaN          NaN          NaN   
227       Syrian Arab Republic    51.794024    63.419625    77.727013   
228   Turks and Caicos Islands          NaN          NaN          NaN   
237                Timor-Leste          NaN          NaN          NaN   
255     British Virgin Islands          NaN          NaN          NaN   
256      Virgin Islands (U.S.)          NaN          NaN          NaN   
261                     Kosovo          NaN          NaN          NaN   
262                Yemen, Rep.    65.549442    70.729858    68.381964   
265                   Zimbabwe          NaN          NaN          NaN   

            2003         2004         2005         2006         2007  \
0            NaN          NaN          NaN          NaN          NaN   
2       6.277876     5.199076     5.629746     5.538590     6.339159   
11           NaN          NaN          NaN          NaN          NaN   
27           NaN          NaN          NaN          NaN          NaN   
38           NaN          NaN          NaN          NaN          NaN   
51           NaN          NaN          NaN          NaN          NaN   
52           NaN          NaN          NaN          NaN          NaN   
78           NaN          NaN          NaN          NaN          NaN   
84           NaN          NaN          NaN          NaN          NaN   
91           NaN          NaN          NaN          NaN          NaN   
93           NaN          NaN          NaN          NaN          NaN   
96           NaN          NaN          NaN          NaN          NaN   
108          NaN          NaN          NaN          NaN          NaN   
110          NaN          NaN          NaN          NaN          NaN   
113    82.577399   250.561625   188.999818   162.649677   217.358166   
132   550.826561  

In [39]:
# Cleaning Medical Doctor data
medical_doc.head()

country   1960  1961  1962  1963  1964   1965  1966  1967  1968  ...  \
0        Aruba    NaN   NaN   NaN   NaN   NaN    NaN   NaN   NaN   NaN  ...   
1  Afghanistan  0.035   NaN   NaN   NaN   NaN  0.063   NaN   NaN   NaN  ...   
2       Angola  0.067   NaN   NaN   NaN   NaN  0.076   NaN   NaN   NaN  ...   
3      Albania  0.276   NaN   NaN   NaN   NaN  0.481   NaN   NaN   NaN  ...   
4      Andorra    NaN   NaN   NaN   NaN   NaN    NaN   NaN   NaN   NaN  ...   

    2010   2011   2012   2013   2014   2015   2016   2017  2018  2019  
0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   NaN   NaN  
1  0.236  0.252  0.241  0.285  0.298  0.285  0.278    NaN   NaN   NaN  
2    NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.215   NaN   NaN  
3  1.230  1.220  1.270  1.280    NaN    NaN  1.220    NaN  2.16  1.65  
4    NaN    NaN    NaN    NaN    NaN  3.330    NaN    NaN   NaN   NaN  

[5 rows x 61 columns]

In [41]:
#Selecting year range
medical_doc20 = medical_doc[['country'] + [str(year) for year in range(2000,2020)]]
medical_doc20

country   2000  2001  2002   2003    2004    2005    2006    2007  \
0           Aruba    NaN   NaN   NaN    NaN     NaN     NaN     NaN     NaN   
1     Afghanistan    NaN  0.19   NaN    NaN     NaN     NaN  0.1600  0.1740   
2          Angola    NaN   NaN   NaN    NaN  0.0621     NaN     NaN     NaN   
3         Albania  1.380  1.31  1.17  1.190  1.1900     NaN  1.1800  1.2000   
4         Andorra  2.570  2.60  2.90  3.330  3.1300  3.2300  3.0100  3.0100   
..            ...    ...   ...   ...    ...     ...     ...     ...     ...   
202         Samoa    NaN   NaN   NaN  0.282     NaN  0.2780     NaN     NaN   
203         Yemen    NaN   NaN   NaN    NaN  0.3450     NaN     NaN     NaN   
204  South Africa    NaN   NaN   NaN    NaN  0.7370     NaN     NaN  0.6990   
205        Zambia    NaN   NaN   NaN    NaN  0.1300  0.0545  0.0533     NaN   
206      Zimbabwe  0.126   NaN   NaN    NaN  0.1730  0.1260     NaN  0.0544   

       2008  ...    2010    2011    2012    2013   2014   2015    2016   2017  \
0       NaN  ...     NaN     NaN     NaN     NaN    NaN    NaN     NaN    NaN   
1    0.1740  ...  0.2360  0.2520  0.2410  0.2850  0.298  0.285  0.2780    NaN   
2       NaN  ...     NaN     NaN     NaN     NaN    NaN    NaN     NaN  0.215   
3       NaN  ...  1.2300  1.2200  1.2700  1.2800    NaN    NaN  1.2200    NaN   
4       NaN  ...     NaN     NaN     NaN     NaN    NaN  3.330     NaN    NaN   
..      ...  ...     ...     ...     ...     ...    ...    ...     ...    ...   
202  0.4640  ...  0.3440     NaN     NaN  0.4820    NaN    NaN  0.3440    NaN   
203  0.3160  ...     NaN     NaN     NaN     NaN  0.525    NaN     NaN    NaN   
204  0.6970  ...  0.7210  0.7230  0.7280  0.7420  0.754  0.781  0.7670  0.783   
205  0.0619  ...  0.0614     NaN     NaN     NaN    NaN    NaN  0.0925    NaN   
206  0.0596  ...  0.1270  0.0817  0.0807  0.0834  0.124  0.181  0.1790  0.186   

      2018   2019  
0      NaN    NaN  
1      NaN    NaN  
2      NaN    NaN  
3    2.160  1.650  
4      NaN    NaN  
..     ...    ...  
202    NaN    NaN  
203    NaN    NaN  
204  0.787  0.792  
205    NaN    NaN  
206  0.210    NaN  

[207 rows x 21 columns]

In [45]:
#Identifying missing data

medical_doc20[medical_doc20.isnull().any(axis=1)]


country   2000  2001  2002   2003    2004    2005    2006    2007  \
0           Aruba    NaN   NaN   NaN    NaN     NaN     NaN     NaN     NaN   
1     Afghanistan    NaN  0.19   NaN    NaN     NaN     NaN  0.1600  0.1740   
2          Angola    NaN   NaN   NaN    NaN  0.0621     NaN     NaN     NaN   
3         Albania  1.380  1.31  1.17  1.190  1.1900     NaN  1.1800  1.2000   
4         Andorra  2.570  2.60  2.90  3.330  3.1300  3.2300  3.0100  3.0100   
..            ...    ...   ...   ...    ...     ...     ...     ...     ...   
202         Samoa    NaN   NaN   NaN  0.282     NaN  0.2780     NaN     NaN   
203         Yemen    NaN   NaN   NaN    NaN  0.3450     NaN     NaN     NaN   
204  South Africa    NaN   NaN   NaN    NaN  0.7370     NaN     NaN  0.6990   
205        Zambia    NaN   NaN   NaN    NaN  0.1300  0.0545  0.0533     NaN   
206      Zimbabwe  0.126   NaN   NaN    NaN  0.1730  0.1260     NaN  0.0544   

       2008  ...    2010    2011    2012    2013   2014   2015    2016   2017  \
0       NaN  ...     NaN     NaN     NaN     NaN    NaN    NaN     NaN    NaN   
1    0.1740  ...  0.2360  0.2520  0.2410  0.2850  0.298  0.285  0.2780    NaN   
2       NaN  ...     NaN     NaN     NaN     NaN    NaN    NaN     NaN  0.215   
3       NaN  ...  1.2300  1.2200  1.2700  1.2800    NaN    NaN  1.2200    NaN   
4       NaN  ...     NaN     NaN     NaN     NaN    NaN  3.330     NaN    NaN   
..      ...  ...     ...     ...     ...     ...    ...    ...     ...    ...   
202  0.4640  ...  0.3440     NaN     NaN  0.4820    NaN    NaN  0.3440    NaN   
203  0.3160  ...     NaN     NaN     NaN     NaN  0.525    NaN     NaN    NaN   
204  0.6970  ...  0.7210  0.7230  0.7280  0.7420  0.754  0.781  0.7670  0.783   
205  0.0619  ...  0.0614     NaN     NaN     NaN    NaN    NaN  0.0925    NaN   
206  0.0596  ...  0.1270  0.0817  0.0807  0.0834  0.124  0.181  0.1790  0.186   

      2018   2019  
0      NaN    NaN  
1      NaN    NaN  
2      NaN    NaN  
3    2.160  1.650  
4      NaN    NaN  
..     ...    ...  
202    NaN    NaN  
203    NaN    NaN  
204  0.787  0.792  
205    NaN    NaN  
206  0.210    NaN  

[196 rows x 21 columns]

In [ ]:



### Data Dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

*Note*: if you are unsure of what a feature is, check the source of the data! This can be found in the README.

**To-Do:** *Edit the table below to create your own data dictionary for the datasets you chose.*

|Feature|Type|Dataset|Description|
|---|---|---|---|
|column name|int/float/object|This is an example| 


## Exploratory Data Analysis

Complete the following steps to explore your data. You are welcome to do more EDA than the steps outlined here as you feel necessary:
1. Summary Statistics.
2. Use a **dictionary comprehension** to apply the standard deviation function you create in part 1 to each numeric column in one of your dataframes.  **No loops**.
    - Assign the output to variable `sd` as a dictionary where: 
        - Each column name is now a key 
        - That standard deviation of the column is the value
        - *Example Output :* `{'1801': 3.80, '1802': 3.93, ...}`
3. Investigate trends in the data.
    - Using sorting and/or masking (along with the `.head()` method to avoid printing our entire dataframe), consider questions relevant to your problem statement. Some examples are provided below (but feel free to change these questions for your specific problem):
        - Which states have the highest and lowest populations for the past few decades?
        - Which states have the highest and lowest economic growth over the course of the dataset?
        - What years were there new countries created?
        - Do any countries have higher growth rates of population than their economy?
        - Which countries have had the lowest population growth?
    - **You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [ ]:
#Code:

**To-Do:** *Edit this cell with your findings on trends in the data (step 3 above).*

## Visualize the Data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers. It is important to not only create visualizations, but to **interpret your visualizations** as well.

**Every plot should**:
- Have a title
- Have axis labels
- Have appropriate tick labels
- Text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Have an interpretation to aid understanding

Here is an example of what your plots should look like following the above guidelines. Note that while the content of this example is unrelated, the principles of visualization hold:

![](https://snag.gy/hCBR1U.jpg)
*Interpretation: The above image shows that as we increase our spending on advertising, our sales numbers also tend to increase. There is a positive correlation between advertising spending and sales.*

---

Here are some prompts to get you started with visualizations. Feel free to add additional visualizations as you see fit:
1. Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features.
    - Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).
    - Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).
2. Visualize distributions using histograms. If you have a lot, consider writing a custom function and use subplots.
    - *OPTIONAL*: Summarize the underlying distributions of your features (in words & statistics)
         - Be thorough in your verbal description of these distributions.
         - Be sure to back up these summaries with statistics.
         - We generally assume that data we sample from a population will be normally distributed. Do we observe this trend? Explain your answers for each distribution and how you think this will affect estimates made from these data.
3. Plot and interpret boxplots. 
    - Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.
    - Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).
    - Each boxplot should:
        - Only include variables of a similar scale
        - Have clear labels for each variable
        - Have appropriate titles and labels
4. Plot and interpret scatter plots to view relationships between features. Feel free to write a custom function, and subplot if you'd like. Functions save both time and space.
    - Your plots should have:
        - Two clearly labeled axes
        - A proper title
        - Colors and symbols that are clear and unmistakable
5. Additional plots of your choosing.
    - Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

In [ ]:
# Code

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Make sure to answer your question of interest or address your problem statement here.

**To-Do:** *Edit this cell with your conclusions and recommendations.*

**Don't forget to create your README!**

**To-Do:** *If you combine your problem statement, data dictionary, brief summary of your analysis, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.* Don't forget to cite your data sources!